src/man/ecpg.1
.TH ECPG UNIX 11/28/98 PostgreSQL \fIPostgreSQL\fP
.SH NAME
ecpg - embedded SQL preprocessor for C / PostgreSQL
.SH SYNOPSIS
.\" \fBecpg\fR [-v ] [-t] [-I include-path ] [-o outfile ]  file1 [ file2 ] [ ... ]
\fBecpg\fR [-v ] [-t] [-I include-path ] [-o outfile ]  file1 [ file2 ] [ ... ]
.SH DESCRIPTION
.B \fIecpg\fP
is an embedded SQL preprocessor for C / PostgreSQL. It
enables development of C programs with embedded SQL code.
.PP
.B \fIecpg\fP
is ultimately intended to be as compliant as possible with the         
ANSI SQL-2 standard and existing commercial ESQL/C packages.                                               
.SH OPTIONS
.B \fIecpg\fP
interprets the following flags when it is invoked               
on the command line:
.PP
.PD 0
.TP 10
.BI \-v 
Print version information. 
.PD
.TP
.B \-t
Turn off auto-transactin mode.
.PD
.TP
.PD
.TP
.B \-I include-path
Specify additional include path. Defaults are \.,
/usr/local/include, the PostgreSQL include path which is defined at compile
time (default: /usr/local/pgsql/lib), /usr/include
.PD
.TP
.B \-o
Specifies that ecpg should write all its output to outfile.
If no such option is given the output is written to foo.c
(if the input file was named foo.pgc.)
If the input file was named foo.bar the output file will be
named foo.bar.c. 
.PD
.TP
.B file1, file2...
The files to be processed.
.\" 
.SH INSTALLATION
The
.B \fIecpg\fP
preprocessor is built during the PostgreSQL installation.  Binaries and
libraries are installed into the PGBASE (i.e., /usr/local/pgsql/... ) 
subdirectories.
.SH PREPROCESSING FOR COMPILATION
.B \fIecpg\fP
.\" (-d ) (-o file) file.pgc ( 2> ecpf.log)
(-o file) file.pgc 
.LP
.\" The optional \-d flag turns on debugging and 2> ecpg.log
.\" redirects the debug output.  The .pgc extension is an 
.\" arbitrary means of denoting ecpg source.
The .pgc extension is an arbitrary means of denoting ecpg source.
.SH COMPILING AND LINKING
Assuming the \fIPostgreSQL\fP binaries are in /usr/local/pgsql:
.LP
gcc -g -i /usr/local/pgsql/include (-o file) file.c 
-L /usr/local/pgsql/lib -lecpg -lpq
.SH ECPG GRAMMAR
.LP
.SH LIBRARIES
.LP
The preprocessor will prepend two directives to the source:
.LP
\fI#include <ecpgtype.h>\fP and \fI#include <ecpglib.h>\fP
.SH VARIABLE DECLARATION  
Variables declared within ecpg source code must be prepended with:
.LP
EXEC SQL BEGIN DECLARE SECTION;  
.LP        
Similarly, variable declaration sections must terminate with:
.LP
EXEC SQL END DECLARE SECTION;
.LP        
NOTE: prior to version 2.1.0, each variable had to be declared 
on a separate line.  As of version 2.1.0 multiple variables may
be declared on a single line:
.LP
char  foo(16), bar(16);
.LP       
.SH ERROR HANDLING
The SQL communication area is defined with:
.LP
EXEC SQL INCLUDE sqlca;
.LP
NOTE: the lowercase `sqlca'.  While SQL convention may be 
followed, i.e., using uppercase to separate embedded SQL 
from C statements, sqlca (which includes the sqlca.h 
header file) MUST be lowercase.  This is because the EXEC SQL
prefix indicates that this INCLUDE will be parsed by ecpg.
ecpg observes case sensitivity (SQLCA.h will not be found.)
EXEC SQL INCLUDE can be used to include other header files
as long as case sensitivity is observed.
.LP
The sqlprint command is used with the EXEC SQL WHENEVER
statement to turn on error handling throughout the 
program:
.LP
EXEC SQL WHENEVER sqlerror sqlprint;
.LP
EXEC SQL WHENEVER not found sqlprint;
.LP
PLEASE NOTE: this is *not* an exhaustive example of usage for
the EXEC SQL WHENEVER statement.  Further examples of usage may
be found in SQL manuals (e.g., `The LAN TIMES Guide to SQL' by
Groff and Weinberg.)
.LP
.SH CONNECTING TO THE DATABASE SERVER
Prior to version 2.1.0 the database name was single quoted:
.RS
EXEC SQL CONNECT 'test1';
.RE
.LP
As of version 2.1.0, the syntax has been simplified:
.LP
.RS
EXEC SQL CONNECT test1;
.RE
(The database name is no longer quoted.)
.LP
Specifying a server and port name in the connect statement is also possible
as of version 6.4. of PostgreSQL. The syntax is:
.LP
.RS
dbname[@server][:port]
.RE
.LP
or
.LP
.RS
<tcp|unix>:postgresql://server[:port][/dbname][?options]
.RE
.SH QUERIES
.LP
.SS Create Table:
.LP
EXEC SQL CREATE TABLE foo (number int4, ascii char(16));  
.RS
EXEC SQL CREATE UNIQUE index num1 on foo(number); 
.RE
EXEC SQL COMMIT;
.LP 
.SS Insert:
.LP
EXEC SQL INSERT INTO foo (number, ascii)
.RS
VALUES (9999, 'doodad');
.RE
EXEC SQL COMMIT;
.LP
.SS Delete:
.LP
EXEC SQL DELETE FROM foo
.RS
WHERE number = 9999;   
.RE
EXEC SQL COMMIT;
.LP
.SS Singleton Select:
.LP
EXEC SQL SELECT foo INTO :FooBar FROM table1
.RS
WHERE ascii = 'doodad';  
.RE
.LP
.SS Select using Cursors:
.LP
EXEC SQL DECLARE foo_bar CURSOR FOR     
.RS
SELECT number, ascii FROM foo    
.RS
ORDER BY ascii;
.RE
.RE
EXEC SQL FETCH foo_bar INTO :FooBar, DooDad;
.LP
...
EXEC SQL CLOSE foo_bar;
.RS
EXEC SQL COMMIT;
.RE
.LP
.SS Updates
.LP
EXEC SQL UPDATE foo
.RS
SET ascii = 'foobar'
.RE
.RS
WHERE number = 9999;
.RE
EXEC SQL COMMIT;
.LP
.SH BUGS
.LP
The is no EXEC SQL PREPARE statement.
.LP
The complete structure definition MUST be listed
inside the declare section.
.LP
See the TODO file in the source for some more missing features.
.LP
.SH "RETURN VALUE"
.LP
ecpg returns 0 to the shell on successful completion, -1
for errors.
.LP
.SH "SEE ALSO"
.PD 0
.TP
\fIcc\fP(1), \fIpgintro\fP(l), \fIcommit\fP(l), \fIdelete\fP(l)
.TP
\fIfetch\fP(l), \fIselect\fP(l), \fIsql\fP(l) , \fIupdate\fP(l)
.PD
.SH FILES
.PD 0
.TP
.B /usr/src/pgsql/postgresql-${ver}/src/interfaces...
 ./ecpg/include.......source for \fIecpg\fP header files.
 ./ecpg/lib...........source for \fIecpg\fP libraries.
 ./ecpg/preproc.......source for \fIecpg\fP header files.
 ./ecpg/test..........source for \fIecpg\fP libraries.
 (test contains examples of syntax for ecpg SQL-C.)
.PD
.TP
.B /usr/local/pgsql/bin 
\fIPostgreSQL\fP binaries including \fIecpg\fP.
.PD
.TP
.B /usr/local/pgsql/include 
\fIPostgreSQL\fP headers including \fIecpglib.h\fP \fIecpgtype.h\fP 
and \fIsqlca.h\fP.
.PD
.TP
.B /usr/local/pgsql/lib 
\fIPostgreSQL\fP libraries including \fIlibecpg.a\fP and 
\fIlibecpg.so\fP.
.SH AUTHORS
Linus Tolke \fI<linus@epact.se>\fP
- original author of ECPG (up to version 0.2).
.br
.PP
Michael Meskes \fI<meskes@debian.org>\fP
- actual author and maintainer of ECPG.
.br
.PP
Thomas Good \fI<tomg@q8.nrnet.org>\fP
- author of this revision of the ecpg man page.
.br
.zZ

HISTORY
                                 Release Notes

                                  Release 7.3

     Release date: 2002-11-??

Overview

   Major changes in this release:

   Schemas

           Schemas allow users to create objects in their own namespace
           so two people or applications can have tables with the same
           name. There is also a public schema for shared tables.
           Table/index creation can be restricted by removing
           permissions on the public schema.

   Drop Column

           PostgreSQL now supports ALTER TABLE ... DROP COLUMN functionality.

   Table Functions

           Functions returning multiple rows and/or multiple columns are
           now much easier to use than before.  You can call such a
           "table function" in the SELECT FROM clause, treating its output
           like a table. Also, plpgsql functions can now return sets.

   Prepared Queries

           For performance, PostgreSQL now supports prepared queries.

   Dependency Tracking

           PostgreSQL now records object dependencies, which allows
           improvements in many areas.

   Privileges

           Functions and procedural languages now have privileges, and
           people running them can take on the privileges of their creators.

   Multibyte/Locale

           Both multibyte and locale are now always enabled.

   Logging

           A variety of logging options have been enhanced.

   Interfaces

           A large number of interfaces have been moved to 
           http://gborg.postgresql.org where they can be developed
           and released independently.

   Functions/Identifiers

           By default, functions can now take up to 32 parameters, and 
           identifiers can be up to 63 bytes long.

     ----------------------------------------------------------------------

Migration to version 7.3

   A dump/restore using "pg_dump" is required for those wishing to migrate
   data from any previous release.  If your application examines the
   system catalogs, additional changes will be required due to the
   introduction of schemas in 7.3; for more information, see:

		http://www.ca.postgresql.org/docs/momjian/upgrade_tips_7.3

   Observe the following incompatibilities:

     * Pre-6.3 clients are no longer supported.

     * pg_hba.conf now has a USER column and additional features.

     * Several GUC logging parameters have been renamed and improved.

     * LIMIT #,# is disabled, use LIMIT # OFFSET #.

     * INSERTs with column lists must specify all values, e.g. 
       INSERT INTO tab (col1, col2) VALUES ('val1') is now invalid.

     * DROP object now takes either CASCADE or RESTRICT to control
       whether dependent objects are also dropped.

     * An index is now not automatically created for SERIAL columns.

     * A SET inside an aborted transaction is now rolled back.

     * COPY no longer considers missing trailing columns to be NULL.

     * TIMESTAMP and TIME data types now default to WITHOUT TIMEZONE.

     * Pre-7.3 databases loaded into 7.3 will not have the new object
       dependencies for SERIAL, UNIQUE constraints, and foreign keys. See
       /contrib/adddepend for a detailed description and a script that
       will add the such dependencies.

     ----------------------------------------------------------------------

Changes

2002-10-23

Server Operation
================
Kerberos V Heimdal support (KTH)
Add pg_locks view to show locks (Neil)
Security fixes for password negotiation memory allocation (Neil)
Remove support for version 0, <=6.2 FE/BE protocol (Tom)
Reserve the last few backend slots for superusers, add GUC variable
  superuser_reserved_connections to control this (Nigel J. Andrews)

Performance
===========
Improve startup by calling localtime() only once (Tom)
Cache system catalog information in flat files for faster startup (Tom)
Improve caching of index information (Tom)
Optimizer improvements (Tom, Fernando Nasser)
Catalog caches now store failed lookups (Tom)
Hash function improvements (Neil)
Improve performance of query tokenization and network handling (Peter)
Speed improvement for large object restore (Mario Weilguni)
Mark expired index entries on first lookup, saving later heap fetches (Tom)
Eliminate NULL bitmap padding when not required (Manfred)
Add BSD-licensed qsort() for Solaris, for performance (Bruce)
Reduce per-row overhead by four bytes (Manfred Koizar)
Fix GEQO optimizer bug (Neil Conway)
Make WITHOUT OID actually save four bytes per row (Manfred Koizar)
Add GUC default_statistics_target variable to specify ANALYZE buckets (Neil)
Use local buffer cache for temporary tables so no WAL overhead (Tom)
Improve free space map performance on large tables (Stephen Marshall, Tom)
Improved WAL write concurrency (Tom)

Privileges
==========
Add privileges on functions and procedural languages (Peter)
Add OWNER to CREATE DATABASE so superusers can create databases
  on behalf of unprivileged users.  (Gavin Sherry, Tom)
Add new object permission bits EXECUTE and USAGE (Tom)
Add SET SESSION AUTHORIZATION DEFAULT and RESET SESSION AUTHORIZATION (Peter)
Allow functions to be executed with the privilege of the function owner (Peter)

Client Authentication
=====================
Allow $HOME/.pgpass to store host/user password combinations (Alvaro Herrera)

Server Configuration
====================
Server log messages now tagged with LOG, not DEBUG (Bruce)
Add USER column to pg_hba.conf (Bruce)
Have log_connections output two lines in log file (Tom)
Remove debug_level from postgresql.conf, now SERVER_MIN_MESSAGES (Bruce)
New ALTER DATABASE/USER ... SET command for per-user/db initialization (Peter)
New GUC variables server_min_messages and client_min_messages which control
  which messages are sent to the server logs or client applications (Bruce)
Allow pg_hba.conf to specify lists of users/dbs separated by commas, 
  group names prepended with +, and file names prepended with @ (Bruce)

src/interfaces/ecpg/preproc/ecpg.c
/* src/interfaces/ecpg/preproc/ecpg.c */

/* Main for ecpg, the PostgreSQL embedded SQL precompiler. */
/* Copyright (c) 1996-2013, PostgreSQL Global Development Group */

#include "postgres_fe.h"

#include <unistd.h>
#include <string.h>
#include "getopt_long.h"

#include "extern.h"

int			ret_value = 0;
bool		autocommit = false,
			auto_create_c = false,
			system_includes = false,
			force_indicator = true,
			questionmarks = false,
			regression_mode = false,
			auto_prepare = false;

char	   *output_filename;

enum COMPAT_MODE compat = ECPG_COMPAT_PGSQL;

struct _include_path *include_paths = NULL;
struct cursor *cur = NULL;
struct typedefs *types = NULL;
struct _defines *defines = NULL;

static void
help(const char *progname)
{
	printf(_("%s is the PostgreSQL embedded SQL preprocessor for C programs.\n\n"),
		   progname);
	printf(_("Usage:\n"
			 "  %s [OPTION]... FILE...\n\n"),
		   progname);
	printf(_("Options:\n"));
	printf(_("  -c             automatically generate C code from embedded SQL code;\n"
			 "                 this affects EXEC SQL TYPE\n"));
	printf(_("  -C MODE        set compatibility mode; MODE can be one of\n"
			 "                 \"INFORMIX\", \"INFORMIX_SE\"\n"));
#ifdef YYDEBUG
	printf(_("  -d             generate parser debug output\n"));
#endif
	printf(_("  -D SYMBOL      define SYMBOL\n"));
	printf(_("  -h             parse a header file, this option includes option \"-c\"\n"));
	printf(_("  -i             parse system include files as well\n"));
	printf(_("  -I DIRECTORY   search DIRECTORY for include files\n"));
	printf(_("  -o OUTFILE     write result to OUTFILE\n"));
	printf(_("  -r OPTION      specify run-time behavior; OPTION can be:\n"
	 "                 \"no_indicator\", \"prepare\", \"questionmarks\"\n"));
	printf(_("  --regression   run in regression testing mode\n"));
	printf(_("  -t             turn on autocommit of transactions\n"));
	printf(_("  -V, --version  output version information, then exit\n"));
	printf(_("  -?, --help     show this help, then exit\n"));
	printf(_("\nIf no output file is specified, the name is formed by adding .c to the\n"
			 "input file name, after stripping off .pgc if present.\n"));
	printf(_("\nReport bugs to <pgsql-bugs@postgresql.org>.\n"));
}

static void
add_include_path(char *path)
{
	struct _include_path *ip = include_paths,
			   *new;

	new = mm_alloc(sizeof(struct _include_path));
	new->path = path;
	new->next = NULL;

	if (ip == NULL)
		include_paths = new;
	else
	{
		for (; ip->next != NULL; ip = ip->next);
		ip->next = new;
	}
}

static void
add_preprocessor_define(char *define)
{
	struct _defines *pd = defines;
	char	   *ptr,
			   *define_copy = mm_strdup(define);

	defines = mm_alloc(sizeof(struct _defines));

	/* look for = sign */
	ptr = strchr(define_copy, '=');
	if (ptr != NULL)
	{
		char	   *tmp;

		/* symbol has a value */
		for (tmp = ptr - 1; *tmp == ' '; tmp--);
		tmp[1] = '\0';
		defines->old = define_copy;
		defines->new = ptr + 1;
	}
	else
	{
		defines->old = define_copy;
		defines->new = mm_strdup("1");
	}
	defines->pertinent = true;
	defines->used = NULL;
	defines->next = pd;
}

#define ECPG_GETOPT_LONG_REGRESSION		1
int
main(int argc, char *const argv[])
{
	static struct option ecpg_options[] = {
		{"regression", no_argument, NULL, ECPG_GETOPT_LONG_REGRESSION},
		{NULL, 0, NULL, 0}
	};

	int			fnr,
				c,
				out_option = 0;
	bool		verbose = false,
				header_mode = false;
	struct _include_path *ip;
	const char *progname;
	char		my_exec_path[MAXPGPATH];
	char		include_path[MAXPGPATH];

	set_pglocale_pgservice(argv[0], PG_TEXTDOMAIN("ecpg"));

	progname = get_progname(argv[0]);

	find_my_exec(argv[0], my_exec_path);

	if (argc > 1)
	{
		if (strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-?") == 0)
		{
			help(progname);
			exit(0);
		}
		if (strcmp(argv[1], "--version") == 0 || strcmp(argv[1], "-V") == 0)
		{
			printf("ecpg (PostgreSQL %s) %d.%d.%d\n", PG_VERSION,
				   MAJOR_VERSION, MINOR_VERSION, PATCHLEVEL);
			exit(0);
		}
	}

	output_filename = NULL;
	while ((c = getopt_long(argc, argv, "vcio:I:tD:dC:r:h", ecpg_options, NULL)) != -1)
	{
		switch (c)
		{
			case ECPG_GETOPT_LONG_REGRESSION:
				regression_mode = true;
				break;
			case 'o':
				output_filename = strdup(optarg);
				if (strcmp(output_filename, "-") == 0)
					base_yyout = stdout;
				else
					base_yyout = fopen(output_filename, PG_BINARY_W);

				if (base_yyout == NULL)
				{
					fprintf(stderr, _("%s: could not open file \"%s\": %s\n"),
							progname, output_filename, strerror(errno));
					output_filename = NULL;
				}
				else
					out_option = 1;
				break;
			case 'I':
				add_include_path(optarg);
				break;
			case 't':
				autocommit = true;
				break;
			case 'v':
				verbose = true;
				break;
			case 'h':
				header_mode = true;
				/* this must include "-c" to make sense */
				/* so do not place a "break;" here */
			case 'c':
				auto_create_c = true;
				break;
			case 'i':
				system_includes = true;
				break;
			case 'C':
				if (strncmp(optarg, "INFORMIX", strlen("INFORMIX")) == 0)
				{
					char		pkginclude_path[MAXPGPATH];
					char		informix_path[MAXPGPATH];

					compat = (strcmp(optarg, "INFORMIX") == 0) ? ECPG_COMPAT_INFORMIX : ECPG_COMPAT_INFORMIX_SE;
					get_pkginclude_path(my_exec_path, pkginclude_path);
					snprintf(informix_path, MAXPGPATH, "%s/informix/esql", pkginclude_path);
					add_include_path(informix_path);
				}
				else
				{
					fprintf(stderr, _("Try \"%s --help\" for more information.\n"), argv[0]);
					return ILLEGAL_OPTION;
				}
				break;
			case 'r':
				if (strcmp(optarg, "no_indicator") == 0)
					force_indicator = false;
				else if (strcmp(optarg, "prepare") == 0)
					auto_prepare = true;
				else if (strcmp(optarg, "questionmarks") == 0)
					questionmarks = true;
				else
				{
					fprintf(stderr, _("Try \"%s --help\" for more information.\n"), argv[0]);
					return ILLEGAL_OPTION;
				}
				break;
			case 'D':
				add_preprocessor_define(optarg);
				break;
			case 'd':
#ifdef YYDEBUG
				base_yydebug = 1;
#else
				fprintf(stderr, _("%s: parser debug support (-d) not available\n"),
						progname);
#endif
				break;
			default:
				fprintf(stderr, _("Try \"%s --help\" for more information.\n"), argv[0]);
				return ILLEGAL_OPTION;
		}
	}

	add_include_path(".");
	add_include_path("/usr/local/include");
	get_include_path(my_exec_path, include_path);
	add_include_path(include_path);
	add_include_path("/usr/include");

	if (verbose)
	{
		fprintf(stderr, _("%s, the PostgreSQL embedded C preprocessor, version %d.%d.%d\n"),
				progname, MAJOR_VERSION, MINOR_VERSION, PATCHLEVEL);
		fprintf(stderr, _("EXEC SQL INCLUDE ... search starts here:\n"));
		for (ip = include_paths; ip != NULL; ip = ip->next)
			fprintf(stderr, " %s\n", ip->path);
		fprintf(stderr, _("end of search list\n"));
		return 0;
	}

	if (optind >= argc)			/* no files specified */
	{
		fprintf(stderr, _("%s: no input files specified\n"), progname);
		fprintf(stderr, _("Try \"%s --help\" for more information.\n"), argv[0]);
		return (ILLEGAL_OPTION);
	}
	else
	{
		/* after the options there must not be anything but filenames */
		for (fnr = optind; fnr < argc; fnr++)
		{
			char	   *ptr2ext;

			/* If argv[fnr] is "-" we have to read from stdin */
			if (strcmp(argv[fnr], "-") == 0)
			{
				input_filename = mm_alloc(strlen("stdin") + 1);
				strcpy(input_filename, "stdin");
				base_yyin = stdin;
			}
			else
			{
				input_filename = mm_alloc(strlen(argv[fnr]) + 5);
				strcpy(input_filename, argv[fnr]);

				/* take care of relative paths */
				ptr2ext = last_dir_separator(input_filename);
				ptr2ext = (ptr2ext ? strrchr(ptr2ext, '.') : strrchr(input_filename, '.'));

				/* no extension? */
				if (ptr2ext == NULL)
				{
					ptr2ext = input_filename + strlen(input_filename);

					/* no extension => add .pgc or .pgh */
					ptr2ext[0] = '.';
					ptr2ext[1] = 'p';
					ptr2ext[2] = 'g';
					ptr2ext[3] = (header_mode == true) ? 'h' : 'c';
					ptr2ext[4] = '\0';
				}

				base_yyin = fopen(input_filename, PG_BINARY_R);
			}

			if (out_option == 0)	/* calculate the output name */
			{
				if (strcmp(input_filename, "stdin") == 0)
					base_yyout = stdout;
				else
				{
					output_filename = mm_alloc(strlen(input_filename) + 3);
					strcpy(output_filename, input_filename);

					ptr2ext = strrchr(output_filename, '.');
					/* make extension = .c resp. .h */
					ptr2ext[1] = (header_mode == true) ? 'h' : 'c';
					ptr2ext[2] = '\0';

					base_yyout = fopen(output_filename, PG_BINARY_W);
					if (base_yyout == NULL)
					{
						fprintf(stderr, _("%s: could not open file \"%s\": %s\n"),
								progname, output_filename, strerror(errno));
						free(output_filename);
						output_filename = NULL;
						free(input_filename);
						continue;
					}
				}
			}

			if (base_yyin == NULL)
				fprintf(stderr, _("%s: could not open file \"%s\": %s\n"),
						progname, argv[fnr], strerror(errno));
			else
			{
				struct cursor *ptr;
				struct _defines *defptr;
				struct typedefs *typeptr;

				/* remove old cursor definitions if any are still there */
				for (ptr = cur; ptr != NULL;)
				{
					struct cursor *this = ptr;
					struct arguments *l1,
							   *l2;

					free(ptr->command);
					free(ptr->connection);
					free(ptr->name);
					for (l1 = ptr->argsinsert; l1; l1 = l2)
					{
						l2 = l1->next;
						free(l1);
					}
					for (l1 = ptr->argsresult; l1; l1 = l2)
					{
						l2 = l1->next;
						free(l1);
					}
					ptr = ptr->next;
					free(this);
				}
				cur = NULL;

				/* remove non-pertinent old defines as well */
				while (defines && !defines->pertinent)
				{
					defptr = defines;
					defines = defines->next;

					free(defptr->new);
					free(defptr->old);
					free(defptr);
				}

				for (defptr = defines; defptr != NULL; defptr = defptr->next)
				{
					struct _defines *this = defptr->next;

					if (this && !this->pertinent)
					{
						defptr->next = this->next;

						free(this->new);
						free(this->old);
						free(this);
					}
				}

				/* and old typedefs */
				for (typeptr = types; typeptr != NULL;)
				{
					struct typedefs *this = typeptr;

					free(typeptr->name);
					ECPGfree_struct_member(typeptr->struct_member_list);
					free(typeptr->type);
					typeptr = typeptr->next;
					free(this);
				}
				types = NULL;

				/* initialize whenever structures */
				memset(&when_error, 0, sizeof(struct when));
				memset(&when_nf, 0, sizeof(struct when));
				memset(&when_warn, 0, sizeof(struct when));

				/* and structure member lists */
				memset(struct_member_list, 0, sizeof(struct_member_list));

				/*
				 * and our variable counter for out of scope cursors'
				 * variables
				 */
				ecpg_internal_var = 0;

				/* finally the actual connection */
				connection = NULL;

				/* initialize lex */
				lex_init();

				/* we need several includes */
				/* but not if we are in header mode */
				if (regression_mode)
					fprintf(base_yyout, "/* Processed by ecpg (regression mode) */\n");
				else
					fprintf(base_yyout, "/* Processed by ecpg (%d.%d.%d) */\n", MAJOR_VERSION, MINOR_VERSION, PATCHLEVEL);

				if (header_mode == false)
				{
					fprintf(base_yyout, "/* These include files are added by the preprocessor */\n#include <ecpglib.h>\n#include <ecpgerrno.h>\n#include <sqlca.h>\n");

					/* add some compatibility headers */
					if (INFORMIX_MODE)
						fprintf(base_yyout, "/* Needed for informix compatibility */\n#include <ecpg_informix.h>\n");

					fprintf(base_yyout, "/* End of automatic include section */\n");
				}

				if (regression_mode)
					fprintf(base_yyout, "#define ECPGdebug(X,Y) ECPGdebug((X)+100,(Y))\n");

				output_line_number();

				/* and parse the source */
				base_yyparse();

				/*
				 * Check whether all cursors were indeed opened.  It does not
				 * really make sense to declare a cursor but not open it.
				 */
				for (ptr = cur; ptr != NULL; ptr = ptr->next)
					if (!(ptr->opened))
						mmerror(PARSE_ERROR, ET_WARNING, "cursor \"%s\" has been declared but not opened", ptr->name);

				if (base_yyin != NULL && base_yyin != stdin)
					fclose(base_yyin);
				if (out_option == 0 && base_yyout != stdout)
					fclose(base_yyout);

				/*
				 * If there was an error, delete the output file.
				 */
				if (ret_value != 0)
				{
					if (strcmp(output_filename, "-") != 0 && unlink(output_filename) != 0)
						fprintf(stderr, _("could not remove output file \"%s\"\n"), output_filename);
				}
			}

			if (output_filename && out_option == 0) {
				free(output_filename);
				output_filename = NULL;
			}

			free(input_filename);
		}
	}
	return ret_value;
}
